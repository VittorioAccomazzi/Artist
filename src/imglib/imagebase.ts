import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}