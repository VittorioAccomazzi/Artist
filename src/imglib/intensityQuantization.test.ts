import {hash, getCanvases, toSeqCanvas, dumpImage, dumpCanvas, toNodeCanvas, toCanvas} from './testutils'
import {ImageFactory} from './imagebase'
import IntensityQuantization from './intensityQuantization'

test('shall quantize the image', async ()=>{

    let width = 320
    let height= 281
    let int1 = 100
    let int2 = 250
    let xCenter = 160
    let yCenter = 140
    let radius = 100

    let image = ImageFactory.Uint8(width,height)

    // first intensity
    image.imagePixels.forEach((v,i)=>image.imagePixels[i]=int1)

    // second intensities
    for(let y=0; y<height; y++){
        for(let x=0; x<width; x++){
            let dx = x-xCenter 
            let dy = y-yCenter
            let dst= Math.sqrt( dx*dx+dy*dy)
            if( dst < radius ){
                let w = dst/radius
                w *= w // on prpose do not make it linear
                let int = Math.floor(w * int1 + (1-w)*int2 )
                image.set(x,y,int)
            }
        }
    }

    let img0 = image.convertTo(image.imageType)
    let img1 = image.convertTo(image.imageType)
    let img2 = image.convertTo(image.imageType)

    //await dumpImage(image,`gradient image`)

    IntensityQuantization.Run(img0,15,0)
    //await dumpImage(img0,`gradient image quantized 0`)
    let hsh0 = await hash(img0)
    expect(hsh0).toMatchSnapshot()

    IntensityQuantization.Run(img1,15,1)
    //await dumpImage(img1,`gradient image quantized 1`)
    let hsh1 = await hash(img1)
    expect(hsh1).toMatchSnapshot()

    IntensityQuantization.Run(img2,15,2)
    //await dumpImage(img2,`gradient image quantized 2`)
    let hsh2 = await hash(img2)
    expect(hsh2).toMatchSnapshot()

})