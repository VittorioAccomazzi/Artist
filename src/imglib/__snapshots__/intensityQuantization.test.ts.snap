// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall quantize the image 1`] = `"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007ffe000000000000000000000000000ffffff00000000000000000000000003ffffffc000000000000000000000000ffffffff000000000000000000000003ffffffffc00000000000000000000007fffffffff0000000000000000000001ffffffffffc000000000000000000003ffffffffffe00000000000000000000ffffffffffff00000000000000000001ffffffffffff80000000000000000003ffffffffffffc0000000000000000007ffffffffffffe000000000000000000ffffffffffffff00000000000000000003ffffffffffe000000000000000000007fffffffffff00000000000000000000ffffffffffff00000000000000000001ffffffffffffc0000000000000000003ffffffffffffc0000000000000000007ffffffffffffe000000000000000000ffffffffffffff000000000000000000ffffffffffffff000000000000000001ffffffffffffff800000000000000003ffffffffffffffc00000000000000003ffffffffffffffc00000000000000003ffffffffffffffe00000000000000007ffffffffffffffe00000000000000007fffffffffffffff0000000000000000ffffffffffffffff0000000000000000ffffffffffffffff8000000000000001ffffffffffffffff8000000000000001ffffffffffffffff8000000000000001ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffe000000000000003ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000003ffffffffffffffffe000000000000003ffffffffffffffffe000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000001ffffffffffffffffc000000000000001ffffffffffffffff8000000000000001ffffffffffffffff8000000000000000ffffffffffffffff8000000000000000ffffffffffffffff00000000000000007fffffffffffffff00000000000000007ffffffffffffffe00000000000000003ffffffffffffffe00000000000000003ffffffffffffffc00000000000000003ffffffffffffffc00000000000000001ffffffffffffff800000000000000000ffffffffffffff000000000000000000ffffffffffffff0000000000000000007ffffffffffffe0000000000000000003ffffffffffffc0000000000000000001ffffffffffffc0000000000000000000ffffffffffff000000000000000000007fffffffffff000000000000000000003ffffffffffe0000000000000000000ffffffffffffff0000000000000000007ffffffffffffe0000000000000000003ffffffffffffc0000000000000000001ffffffffffff80000000000000000000ffffffffffff000000000000000000003ffffffffffe000000000000000000001ffffffffffc0000000000000000000007fffffffff00000000000000000000003ffffffffc00000000000000000000000ffffffff0000000000000000000000003ffffffc0000000000000000000000000ffffff0000000000000000000000000007ffe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall quantize the image 2`] = `"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007ffe000000000000000000000000000ffffff00000000000000000000000003ffffffc000000000000000000000000ffffffff000000000000000000000003ffffffffc00000000000000000000007fffffffff0000000000000000000001ffffffffffc000000000000000000003ffffffffffe00000000000000000000ffffffffffff00000000000000000001ffffffffffff80000000000000000003ffffffffffffc0000000000000000007ffffffffffffe000000000000000000ffffffffffffff00000000000000000007ffffffffffe00000000000000000000ffffffffffff00000000000000000001ffffffffffff80000000000000000003ffffffffffffc0000000000000000007ffffffffffffe0000000000000000007fffffffffffff000000000000000000ffffffffffffff800000000000000001ffffffffffffff800000000000000001ffffffffffffffc00000000000000003ffffffffffffffc00000000000000007ffffffffffffffe00000000000000007ffffffffffffffe0000000000000000ffffffffffffffff0000000000000000ffffffffffffffff0000000000000000ffffffffffffffff8000000000000001ffffffffffffffff8000000000000001ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffe000000000000003ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007fffffffffffffffff000000000000007fffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000003ffffffffffffffffe000000000000003ffffffffffffffffe000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000001ffffffffffffffffc000000000000001ffffffffffffffff8000000000000000ffffffffffffffff8000000000000000ffffffffffffffff0000000000000000ffffffffffffffff00000000000000007ffffffffffffffe00000000000000007ffffffffffffffe00000000000000003ffffffffffffffc00000000000000001ffffffffffffffc00000000000000001ffffffffffffff800000000000000000ffffffffffffff8000000000000000007fffffffffffff0000000000000000007ffffffffffffe0000000000000000003ffffffffffffc0000000000000000001ffffffffffff80000000000000000000ffffffffffff000000000000000000007ffffffffffe0000000000000000000ffffffffffffff0000000000000000007ffffffffffffe0000000000000000003ffffffffffffc0000000000000000001ffffffffffff80000000000000000000ffffffffffff000000000000000000003ffffffffffe000000000000000000001ffffffffffc0000000000000000000007fffffffff00000000000000000000003ffffffffc00000000000000000000000ffffffff0000000000000000000000003ffffffc0000000000000000000000000ffffff0000000000000000000000000007ffe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`shall quantize the image 3`] = `"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007ffe000000000000000000000000000ffffff00000000000000000000000003ffffffc000000000000000000000000ffffffff000000000000000000000003ffffffffc00000000000000000000007fffffffff0000000000000000000001ffffffffffc000000000000000000003ffffffffffe00000000000000000000ffffffffffff00000000000000000001ffffffffffff80000000000000000003ffffffffffffc0000000000000000007ffffffffffffe000000000000000000ffffffffffffff00000000000000000007ffffffffffe00000000000000000000ffffffffffff00000000000000000001ffffffffffff80000000000000000003ffffffffffffc0000000000000000007ffffffffffffe0000000000000000007fffffffffffff000000000000000000ffffffffffffff000000000000000001ffffffffffffff800000000000000001ffffffffffffffc00000000000000003ffffffffffffffc00000000000000003ffffffffffffffe00000000000000007fffffffffffffff0000000000000000ffffffffffffffff0000000000000000ffffffffffffffff0000000000000000ffffffffffffffff8000000000000001ffffffffffffffff8000000000000001ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000003ffffffffffffffffe000000000000003ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff00000000000000ffffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007fffffffffffffffff000000000000007ffffffffffffffffe000000000000007ffffffffffffffffe000000000000003ffffffffffffffffe000000000000003ffffffffffffffffe000000000000003ffffffffffffffffc000000000000003ffffffffffffffffc000000000000001ffffffffffffffffc000000000000001ffffffffffffffff8000000000000000ffffffffffffffff8000000000000000ffffffffffffffff0000000000000000ffffffffffffffff00000000000000007fffffffffffffff00000000000000003ffffffffffffffe00000000000000003ffffffffffffffc00000000000000001ffffffffffffffc00000000000000001ffffffffffffff800000000000000000ffffffffffffff0000000000000000007fffffffffffff0000000000000000007ffffffffffffe0000000000000000003ffffffffffffc0000000000000000001ffffffffffff80000000000000000000ffffffffffff000000000000000000007ffffffffffe0000000000000000000ffffffffffffff0000000000000000007ffffffffffffe0000000000000000003ffffffffffffc0000000000000000001ffffffffffff80000000000000000000ffffffffffff000000000000000000003ffffffffffe000000000000000000001ffffffffffc0000000000000000000007fffffffff00000000000000000000003ffffffffc00000000000000000000000ffffffff0000000000000000000000003ffffffc0000000000000000000000000ffffff0000000000000000000000000007ffe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;
